File: scripts/skills_access_diagnosis.py
#!/usr/bin/env python3
from __future__ import annotations

import json
import os
import shutil
import subprocess
from pathlib import Path
from typing import Any

ROOT = Path(__file__).resolve().parent.parent
DEFAULT_WORKSPACE = Path(os.environ.get("OPENCLAW_WORKSPACE", str(ROOT))).expanduser()
HOME = Path.home()
GLOBAL_SKILLS = HOME / ".openclaw" / "skills"
WORKSPACE_SKILLS = DEFAULT_WORKSPACE / "skills"
REPO_SKILLS = ROOT / "skills"


def run(cmd: list[str], timeout: int = 15) -> dict[str, Any]:
    try:
        proc = subprocess.run(cmd, capture_output=True, text=True, timeout=timeout)
        return {
            "ok": proc.returncode == 0,
            "returncode": proc.returncode,
            "stdout": proc.stdout[-4000:],
            "stderr": proc.stderr[-4000:],
        }
    except FileNotFoundError:
        return {"ok": False, "error": "missing-binary"}
    except subprocess.TimeoutExpired:
        return {"ok": False, "error": "timeout", "timeout_seconds": timeout}


def directory_summary(path: Path) -> dict[str, Any]:
    exists = path.exists()
    is_dir = path.is_dir()
    children = []
    if exists and is_dir:
        children = sorted(entry.name for entry in path.iterdir() if entry.is_dir())

    return {
        "path": str(path),
        "exists": exists,
        "is_dir": is_dir,
        "skill_count": len(children),
        "sample": children[:10],
    }


def main() -> None:
    report: dict[str, Any] = {
        "paths": {
            "repo_root": directory_summary(ROOT),
            "repo_skills": directory_summary(REPO_SKILLS),
            "workspace": directory_summary(DEFAULT_WORKSPACE),
            "workspace_skills": directory_summary(WORKSPACE_SKILLS),
            "global_skills": directory_summary(GLOBAL_SKILLS),
            "config": {
                "path": str(HOME / ".openclaw" / "openclaw.json"),
                "exists": (HOME / ".openclaw" / "openclaw.json").exists(),
                "is_dir": (HOME / ".openclaw" / "openclaw.json").is_dir(),
            },
        },
        "skill_search_order": [
            str(WORKSPACE_SKILLS),
            str(GLOBAL_SKILLS),
            "bundled skills",
        ],
        "manual_install_targets": {
            "preferred_default": str(GLOBAL_SKILLS),
            "workspace_override": str(WORKSPACE_SKILLS),
        },
        "binaries": {
            "openclaw": shutil.which("openclaw"),
            "clawhub": shutil.which("clawhub"),
            "jq": shutil.which("jq"),
        },
        "checks": {},
        "recommendations": [
            "Prefer targeted installs over bulk updates during incidents.",
            "Try one targeted 'clawhub install <skill-slug>' first; if it stalls for roughly 30 seconds, stop it and switch to the manual fallback.",
            "If registry install hangs, review the skill source and fall back to 'bash scripts/install-skill-fallback.sh /path/to/skill --global'.",
            "Use '--workspace' only when you intentionally want a repo-scoped override that should win over global or bundled skills.",
            "Restart the agent session after adding or changing skills so the refreshed skill snapshot is picked up.",
        ],
    }

    if report["binaries"]["openclaw"]:
        report["checks"]["openclaw_skills_list"] = run(["openclaw", "skills", "list"])
        report["checks"]["openclaw_skills_eligible"] = run(["openclaw", "skills", "list", "--eligible"])
        report["checks"]["openclaw_skills_check"] = run(["openclaw", "skills", "check"])

    if report["binaries"]["clawhub"]:
        report["checks"]["clawhub_help"] = run(["clawhub", "--help"], timeout=10)

    print(json.dumps(report, indent=2))


if __name__ == "__main__":
    main()

File: scripts/skills-access-diagnosis.mjs
#!/usr/bin/env node

import fs from "node:fs";
import os from "node:os";
import path from "node:path";
import { spawnSync } from "node:child_process";

const root = path.resolve(import.meta.dirname, "..");
const home = os.homedir();
const defaultWorkspace = path.resolve(process.env.OPENCLAW_WORKSPACE || root);
const globalSkills = path.join(home, ".openclaw", "skills");
const workspaceSkills = path.join(defaultWorkspace, "skills");
const repoSkills = path.join(root, "skills");

function run(cmd, args, options = {}) {
	try {
		const result = spawnSync(cmd, args, {
			encoding: "utf8",
			timeout: options.timeout ?? 15000
		});
		if (result.error) {
			if (result.error.code === "ENOENT") return { ok: false, error: "missing-binary" };
			if (result.error.code === "ETIMEDOUT") return { ok: false, error: "timeout", timeout_seconds: (options.timeout ?? 15000) / 1000 };
			return { ok: false, error: String(result.error.message || result.error) };
		}
		return {
			ok: result.status === 0,
			returncode: result.status ?? 1,
			stdout: (result.stdout || "").slice(-4000),
			stderr: (result.stderr || "").slice(-4000)
		};
	} catch (error) {
		return { ok: false, error: String(error?.message || error) };
	}
}

function directorySummary(targetPath) {
	const exists = fs.existsSync(targetPath);
	const isDir = exists && fs.statSync(targetPath).isDirectory();
	const children = isDir
		? fs
			.readdirSync(targetPath, { withFileTypes: true })
			.filter((entry) => entry.isDirectory())
			.map((entry) => entry.name)
			.sort()
		: [];

	return {
		path: targetPath,
		exists,
		is_dir: isDir,
		skill_count: children.length,
		sample: children.slice(0, 10)
	};
}

function configSummary() {
	const configPath = path.join(home, ".openclaw", "openclaw.json");
	const exists = fs.existsSync(configPath);
	return {
		path: configPath,
		exists,
		is_dir: exists ? fs.statSync(configPath).isDirectory() : false
	};
}

function which(binary) {
	const lookup = process.platform === "win32" ? ["where", binary] : ["which", binary];
	const result = run(lookup[0], lookup.slice(1), { timeout: 10000 });
	if (!result.ok || !result.stdout.trim()) return null;
	return result.stdout.trim().split(/\r?\n/)[0] || null;
}

const report = {
	paths: {
		repo_root: directorySummary(root),
		repo_skills: directorySummary(repoSkills),
		workspace: directorySummary(defaultWorkspace),
		workspace_skills: directorySummary(workspaceSkills),
		global_skills: directorySummary(globalSkills),
		config: configSummary()
	},
	skill_search_order: [workspaceSkills, globalSkills, "bundled skills"],
	manual_install_targets: {
		preferred_default: globalSkills,
		workspace_override: workspaceSkills
	},
	binaries: {
		openclaw: which("openclaw"),
		clawhub: which("clawhub"),
		jq: which("jq")
	},
	checks: {},
	recommendations: [
		"Prefer targeted installs over bulk updates during incidents.",
		"Try one targeted 'clawhub install <skill-slug>' first; if it stalls for roughly 30 seconds, stop it and switch to the manual fallback.",
		"If registry install hangs, review the skill source and fall back to 'bash scripts/install-skill-fallback.sh /path/to/skill --global'.",
		"Use '--workspace' only when you intentionally want a repo-scoped override that should win over global or bundled skills.",
		"Restart the agent session after adding or changing skills so the refreshed skill snapshot is picked up."
	]
};

if (report.binaries.openclaw) {
	report.checks.openclaw_skills_list = run("openclaw", ["skills", "list"]);
	report.checks.openclaw_skills_eligible = run("openclaw", ["skills", "list", "--eligible"]);
	report.checks.openclaw_skills_check = run("openclaw", ["skills", "check"]);
}

if (report.binaries.clawhub) {
	report.checks.clawhub_help = run("clawhub", ["--help"], { timeout: 10000 });
}

console.log(JSON.stringify(report, null, 2));
